{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Ricardo\\AppData\\Local\\Temp\\ipykernel_11768\\3038891008.py:9: DtypeWarning: Columns (9,11,12,13,23,26) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(archivo_entrada_tsv, sep='\\t')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  DATE ADDED TO CATALOG  PUBMEDID   FIRST AUTHOR        DATE         JOURNAL  \\\n",
      "0            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "1            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "2            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "3            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "4            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "\n",
      "                                   LINK  \\\n",
      "0  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "1  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "2  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "3  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "4  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "\n",
      "                                               STUDY  \\\n",
      "0  Genome-wide Association Study Identifies 27 Lo...   \n",
      "1  Genome-wide Association Study Identifies 27 Lo...   \n",
      "2  Genome-wide Association Study Identifies 27 Lo...   \n",
      "3  Genome-wide Association Study Identifies 27 Lo...   \n",
      "4  Genome-wide Association Study Identifies 27 Lo...   \n",
      "\n",
      "                                  DISEASE/TRAIT  \\\n",
      "0  Granulocyte-colony stimulating factor levels   \n",
      "1  Granulocyte-colony stimulating factor levels   \n",
      "2         Growth-regulated protein alpha levels   \n",
      "3         Growth-regulated protein alpha levels   \n",
      "4         Growth-regulated protein alpha levels   \n",
      "\n",
      "                  INITIAL SAMPLE SIZE REPLICATION SAMPLE SIZE  ...  \\\n",
      "0  7,904 Finnish ancestry individuals                     NaN  ...   \n",
      "1  7,904 Finnish ancestry individuals                     NaN  ...   \n",
      "2  3,505 Finnish ancestry individuals                     NaN  ...   \n",
      "3  3,505 Finnish ancestry individuals                     NaN  ...   \n",
      "4  3,505 Finnish ancestry individuals                     NaN  ...   \n",
      "\n",
      "               CONTEXT INTERGENIC RISK ALLELE FREQUENCY       P-VALUE  \\\n",
      "0       intron_variant        0.0                    NR  7.000000e-06   \n",
      "1   intergenic_variant        1.0                    NR  3.000000e-06   \n",
      "2   intergenic_variant        1.0                    NR  5.000000e-29   \n",
      "3  3_prime_UTR_variant        0.0                    NR  3.000000e-21   \n",
      "4       intron_variant        1.0                    NR  2.000000e-09   \n",
      "\n",
      "  PVALUE_MLOG P-VALUE (TEXT) OR or BETA                    95% CI (TEXT)  \\\n",
      "0    5.154902            NaN     0.3273    [0.19-0.47] SD units decrease   \n",
      "1    5.522879            NaN     0.0890  [0.052-0.126] SD units decrease   \n",
      "2   28.301030            NaN     0.3012    [0.25-0.35] SD units increase   \n",
      "3   20.522879            NaN     0.2593    [0.21-0.31] SD units decrease   \n",
      "4    8.698970            NaN     0.1548     [0.1-0.21] SD units decrease   \n",
      "\n",
      "   PLATFORM [SNPS PASSING QC]  CNV  \n",
      "0     NR [10700000] (imputed)    N  \n",
      "1     NR [10700000] (imputed)    N  \n",
      "2     NR [10700000] (imputed)    N  \n",
      "3     NR [10700000] (imputed)    N  \n",
      "4     NR [10700000] (imputed)    N  \n",
      "\n",
      "[5 rows x 34 columns]\n",
      "Index(['DATE ADDED TO CATALOG', 'PUBMEDID', 'FIRST AUTHOR', 'DATE', 'JOURNAL',\n",
      "       'LINK', 'STUDY', 'DISEASE/TRAIT', 'INITIAL SAMPLE SIZE',\n",
      "       'REPLICATION SAMPLE SIZE', 'REGION', 'CHR_ID', 'CHR_POS',\n",
      "       'REPORTED GENE(S)', 'MAPPED_GENE', 'UPSTREAM_GENE_ID',\n",
      "       'DOWNSTREAM_GENE_ID', 'SNP_GENE_IDS', 'UPSTREAM_GENE_DISTANCE',\n",
      "       'DOWNSTREAM_GENE_DISTANCE', 'STRONGEST SNP-RISK ALLELE', 'SNPS',\n",
      "       'MERGED', 'SNP_ID_CURRENT', 'CONTEXT', 'INTERGENIC',\n",
      "       'RISK ALLELE FREQUENCY', 'P-VALUE', 'PVALUE_MLOG', 'P-VALUE (TEXT)',\n",
      "       'OR or BETA', '95% CI (TEXT)', 'PLATFORM [SNPS PASSING QC]', 'CNV'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "import gzip\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "# Define la ruta y el nombre del archivo TSV de entrada\n",
    "archivo_entrada_tsv = 'gwas-catalog-associations.tsv'\n",
    "\n",
    "# Leer el archivo TSV en un DataFrame\n",
    "df = pd.read_csv(archivo_entrada_tsv, sep='\\t')\n",
    "\n",
    "# Mostrar el DataFrame y las columnas para verificar que se haya leído correctamente\n",
    "print(df.head())\n",
    "print(df.columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Columnas completamente llenas de valores nulos: []\n",
      "No se encontraron columnas completamente llenas de nulos para eliminar.\n",
      "  DATE ADDED TO CATALOG  PUBMEDID   FIRST AUTHOR        DATE         JOURNAL  \\\n",
      "0            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "1            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "2            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "3            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "4            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "\n",
      "                                   LINK  \\\n",
      "0  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "1  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "2  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "3  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "4  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "\n",
      "                                               STUDY  \\\n",
      "0  Genome-wide Association Study Identifies 27 Lo...   \n",
      "1  Genome-wide Association Study Identifies 27 Lo...   \n",
      "2  Genome-wide Association Study Identifies 27 Lo...   \n",
      "3  Genome-wide Association Study Identifies 27 Lo...   \n",
      "4  Genome-wide Association Study Identifies 27 Lo...   \n",
      "\n",
      "                                  DISEASE/TRAIT  \\\n",
      "0  Granulocyte-colony stimulating factor levels   \n",
      "1  Granulocyte-colony stimulating factor levels   \n",
      "2         Growth-regulated protein alpha levels   \n",
      "3         Growth-regulated protein alpha levels   \n",
      "4         Growth-regulated protein alpha levels   \n",
      "\n",
      "                  INITIAL SAMPLE SIZE REPLICATION SAMPLE SIZE  ...  \\\n",
      "0  7,904 Finnish ancestry individuals                     NaN  ...   \n",
      "1  7,904 Finnish ancestry individuals                     NaN  ...   \n",
      "2  3,505 Finnish ancestry individuals                     NaN  ...   \n",
      "3  3,505 Finnish ancestry individuals                     NaN  ...   \n",
      "4  3,505 Finnish ancestry individuals                     NaN  ...   \n",
      "\n",
      "               CONTEXT INTERGENIC RISK ALLELE FREQUENCY       P-VALUE  \\\n",
      "0       intron_variant        0.0                    NR  7.000000e-06   \n",
      "1   intergenic_variant        1.0                    NR  3.000000e-06   \n",
      "2   intergenic_variant        1.0                    NR  5.000000e-29   \n",
      "3  3_prime_UTR_variant        0.0                    NR  3.000000e-21   \n",
      "4       intron_variant        1.0                    NR  2.000000e-09   \n",
      "\n",
      "  PVALUE_MLOG P-VALUE (TEXT) OR or BETA                    95% CI (TEXT)  \\\n",
      "0    5.154902            NaN     0.3273    [0.19-0.47] SD units decrease   \n",
      "1    5.522879            NaN     0.0890  [0.052-0.126] SD units decrease   \n",
      "2   28.301030            NaN     0.3012    [0.25-0.35] SD units increase   \n",
      "3   20.522879            NaN     0.2593    [0.21-0.31] SD units decrease   \n",
      "4    8.698970            NaN     0.1548     [0.1-0.21] SD units decrease   \n",
      "\n",
      "   PLATFORM [SNPS PASSING QC]  CNV  \n",
      "0     NR [10700000] (imputed)    N  \n",
      "1     NR [10700000] (imputed)    N  \n",
      "2     NR [10700000] (imputed)    N  \n",
      "3     NR [10700000] (imputed)    N  \n",
      "4     NR [10700000] (imputed)    N  \n",
      "\n",
      "[5 rows x 34 columns]\n"
     ]
    }
   ],
   "source": [
    "# Columnas específicas para verificar si están completamente llenas de valores nulos\n",
    "columnas_especificas = ['DATE ADDED TO CATALOG', 'PUBMEDID', 'FIRST AUTHOR', 'DATE', 'JOURNAL',\n",
    "                        'LINK', 'STUDY', 'DISEASE/TRAIT', 'INITIAL SAMPLE SIZE',\n",
    "                        'REPLICATION SAMPLE SIZE', 'REGION', 'CHR_ID', 'CHR_POS',\n",
    "                        'REPORTED GENE(S)', 'MAPPED_GENE', 'UPSTREAM_GENE_ID',\n",
    "                        'DOWNSTREAM_GENE_ID', 'SNP_GENE_IDS', 'UPSTREAM_GENE_DISTANCE',\n",
    "                        'DOWNSTREAM_GENE_DISTANCE', 'STRONGEST SNP-RISK ALLELE', 'SNPS',\n",
    "                        'MERGED', 'SNP_ID_CURRENT', 'CONTEXT', 'INTERGENIC',\n",
    "                        'RISK ALLELE FREQUENCY', 'P-VALUE', 'PVALUE_MLOG', 'P-VALUE (TEXT)',\n",
    "                        'OR or BETA', '95% CI (TEXT)', 'PLATFORM [SNPS PASSING QC]', 'CNV']\n",
    "\n",
    "# Verificar si las columnas específicas están completamente llenas de valores nulos\n",
    "columnas_con_nulos = [col for col in columnas_especificas if df[col].isnull().all()]\n",
    "\n",
    "# Imprimir las columnas que están completamente llenas de nulos\n",
    "print(\"Columnas completamente llenas de valores nulos:\", columnas_con_nulos)\n",
    "\n",
    "# Si decides eliminar estas columnas específicas, puedes hacerlo así:\n",
    "if columnas_con_nulos:  # Verifica si la lista no está vacía\n",
    "    df = df.drop(columns=columnas_con_nulos)\n",
    "    print(\"Columnas eliminadas. Aquí están las primeras filas del DataFrame actualizado:\")\n",
    "else:\n",
    "    print(\"No se encontraron columnas completamente llenas de nulos para eliminar.\")\n",
    "\n",
    "# Verificar el resultado\n",
    "print(df.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Columnas que serán eliminadas porque contienen al menos un valor nulo: ['INITIAL SAMPLE SIZE', 'REPLICATION SAMPLE SIZE', 'REGION', 'CHR_ID', 'CHR_POS', 'REPORTED GENE(S)', 'MAPPED_GENE', 'UPSTREAM_GENE_ID', 'DOWNSTREAM_GENE_ID', 'SNP_GENE_IDS', 'UPSTREAM_GENE_DISTANCE', 'DOWNSTREAM_GENE_DISTANCE', 'SNP_ID_CURRENT', 'CONTEXT', 'INTERGENIC', 'RISK ALLELE FREQUENCY', 'P-VALUE (TEXT)', 'OR or BETA', '95% CI (TEXT)']\n",
      "  DATE ADDED TO CATALOG  PUBMEDID   FIRST AUTHOR        DATE         JOURNAL  \\\n",
      "0            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "1            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "2            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "3            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "4            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "\n",
      "                                   LINK  \\\n",
      "0  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "1  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "2  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "3  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "4  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "\n",
      "                                               STUDY  \\\n",
      "0  Genome-wide Association Study Identifies 27 Lo...   \n",
      "1  Genome-wide Association Study Identifies 27 Lo...   \n",
      "2  Genome-wide Association Study Identifies 27 Lo...   \n",
      "3  Genome-wide Association Study Identifies 27 Lo...   \n",
      "4  Genome-wide Association Study Identifies 27 Lo...   \n",
      "\n",
      "                                  DISEASE/TRAIT STRONGEST SNP-RISK ALLELE  \\\n",
      "0  Granulocyte-colony stimulating factor levels               rs9785172-C   \n",
      "1  Granulocyte-colony stimulating factor levels               rs1817411-C   \n",
      "2         Growth-regulated protein alpha levels             rs36080575-TA   \n",
      "3         Growth-regulated protein alpha levels               rs4656236-G   \n",
      "4         Growth-regulated protein alpha levels               rs9662662-C   \n",
      "\n",
      "         SNPS  MERGED       P-VALUE  PVALUE_MLOG PLATFORM [SNPS PASSING QC]  \\\n",
      "0   rs9785172       0  7.000000e-06     5.154902    NR [10700000] (imputed)   \n",
      "1   rs1817411       0  3.000000e-06     5.522879    NR [10700000] (imputed)   \n",
      "2  rs36080575       0  5.000000e-29    28.301030    NR [10700000] (imputed)   \n",
      "3   rs4656236       0  3.000000e-21    20.522879    NR [10700000] (imputed)   \n",
      "4   rs9662662       0  2.000000e-09     8.698970    NR [10700000] (imputed)   \n",
      "\n",
      "  CNV  \n",
      "0   N  \n",
      "1   N  \n",
      "2   N  \n",
      "3   N  \n",
      "4   N  \n"
     ]
    }
   ],
   "source": [
    "columnas_con_algun_nulo = df.columns[df.isnull().any()].tolist()\n",
    "\n",
    "print(\"Columnas que serán eliminadas porque contienen al menos un valor nulo:\", columnas_con_algun_nulo)\n",
    "\n",
    "# Eliminar las columnas que contienen al menos un valor nulo\n",
    "df = df.drop(columns=columnas_con_algun_nulo)\n",
    "\n",
    "# Verificar el resultado\n",
    "print(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  DATE ADDED TO CATALOG  PUBMEDID   FIRST AUTHOR        DATE         JOURNAL  \\\n",
      "0            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "1            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "2            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "3            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "4            2017-08-04  27989323  Ahola-Olli AV  2016-12-13  Am J Hum Genet   \n",
      "\n",
      "                                   LINK  \\\n",
      "0  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "1  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "2  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "3  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "4  www.ncbi.nlm.nih.gov/pubmed/27989323   \n",
      "\n",
      "                                               STUDY  \\\n",
      "0  Genome-wide Association Study Identifies 27 Lo...   \n",
      "1  Genome-wide Association Study Identifies 27 Lo...   \n",
      "2  Genome-wide Association Study Identifies 27 Lo...   \n",
      "3  Genome-wide Association Study Identifies 27 Lo...   \n",
      "4  Genome-wide Association Study Identifies 27 Lo...   \n",
      "\n",
      "                                  DISEASE/TRAIT STRONGEST SNP-RISK ALLELE  \\\n",
      "0  Granulocyte-colony stimulating factor levels               rs9785172-C   \n",
      "1  Granulocyte-colony stimulating factor levels               rs1817411-C   \n",
      "2         Growth-regulated protein alpha levels             rs36080575-TA   \n",
      "3         Growth-regulated protein alpha levels               rs4656236-G   \n",
      "4         Growth-regulated protein alpha levels               rs9662662-C   \n",
      "\n",
      "         SNPS  MERGED       P-VALUE  PVALUE_MLOG PLATFORM [SNPS PASSING QC]  \\\n",
      "0   rs9785172       0  7.000000e-06     5.154902    NR [10700000] (imputed)   \n",
      "1   rs1817411       0  3.000000e-06     5.522879    NR [10700000] (imputed)   \n",
      "2  rs36080575       0  5.000000e-29    28.301030    NR [10700000] (imputed)   \n",
      "3   rs4656236       0  3.000000e-21    20.522879    NR [10700000] (imputed)   \n",
      "4   rs9662662       0  2.000000e-09     8.698970    NR [10700000] (imputed)   \n",
      "\n",
      "  CNV  \n",
      "0   N  \n",
      "1   N  \n",
      "2   N  \n",
      "3   N  \n",
      "4   N  \n"
     ]
    }
   ],
   "source": [
    "columnas_sin_nulos = [col for col in columnas_especificas if col not in columnas_con_algun_nulo]\n",
    "\n",
    "# Seleccionar estas columnas en el DataFrame\n",
    "df = df[columnas_sin_nulos]\n",
    "\n",
    "# Verificar el resultado\n",
    "print(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DATE ADDED TO CATALOG                                                2017-08-04\n",
      "PUBMEDID                                                               27989323\n",
      "FIRST AUTHOR                                                      Ahola-Olli AV\n",
      "DATE                                                                 2016-12-13\n",
      "JOURNAL                                                          Am J Hum Genet\n",
      "LINK                                       www.ncbi.nlm.nih.gov/pubmed/27989323\n",
      "STUDY                         Genome-wide Association Study Identifies 27 Lo...\n",
      "DISEASE/TRAIT                             Growth-regulated protein alpha levels\n",
      "STRONGEST SNP-RISK ALLELE                                          rs34197548-C\n",
      "SNPS                                                                 rs34197548\n",
      "MERGED                                                                        0\n",
      "P-VALUE                                                                0.000009\n",
      "PVALUE_MLOG                                                            5.045757\n",
      "PLATFORM [SNPS PASSING QC]                              NR [10700000] (imputed)\n",
      "CNV                                                                           N\n",
      "Name: 5, dtype: object\n"
     ]
    }
   ],
   "source": [
    "specific_row = df.iloc[5]\n",
    "print(specific_row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DataFrame guardado con éxito en 'Salidas\\salida_gwas.tsv'\n"
     ]
    }
   ],
   "source": [
    "# Define la ruta y el nombre del archivo TSV de salida\n",
    "archivo_salida_tsv = 'salida_gwas.tsv'\n",
    "carpeta_salida = 'Salidas'\n",
    "# Asegura que la carpeta de salida exista\n",
    "os.makedirs(carpeta_salida, exist_ok=True)  # Crea la carpeta si no existe\n",
    "# Completar la ruta del archivo de salida\n",
    "archivo_salida_tsv = os.path.join(carpeta_salida, archivo_salida_tsv)\n",
    "\n",
    "# Guardar el DataFrame en un archivo TSV\n",
    "df.to_csv(archivo_salida_tsv, sep='\\t', index=False)\n",
    "\n",
    "# Imprimir confirmación de que el DataFrame ha sido guardado correctamente\n",
    "print(f\"DataFrame guardado con éxito en '{archivo_salida_tsv}'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Advertencia: Las siguientes columnas relevantes no se encontraron en el archivo: ['#chr', 'outermost_start', 'outermost_stop', 'variant', 'method']\n"
     ]
    }
   ],
   "source": [
    "# Lista de columnas relevantes según tu YARRRML\n",
    "# Ajusta esta lista según las columnas que realmente estén definidas en tu YARRRML\n",
    "relevant_columns = ['#chr', 'outermost_start', 'outermost_stop', 'variant', 'method']\n",
    "\n",
    "# Verifica si todas las columnas relevantes existen en el DataFrame, para evitar errores\n",
    "missing_columns = [col for col in relevant_columns if col not in df.columns]\n",
    "if missing_columns:\n",
    "    print(\"Advertencia: Las siguientes columnas relevantes no se encontraron en el archivo:\", missing_columns)\n",
    "else:\n",
    "    # Elimina las columnas que no están en 'relevant_columns'\n",
    "    df = df[relevant_columns]\n",
    "\n",
    "    # Guarda el DataFrame resultante en un nuevo archivo\n",
    "    output_path = 'archivo_filtrado.tsv'  # Asegúrate de actualizar esta ruta\n",
    "    df.to_csv(output_path, sep='\\t', index=False)\n",
    "\n",
    "    print(f\"Archivo guardado exitosamente en: {output_path}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
